### PR TITLE
bpo-38270: More fixes for strict crypto policy

### PR DESCRIPTION
**Pull request by tiran**
test_hmac and test_hashlib test built-in hashing implementations and
OpenSSL-based hashing implementations. Add more checks to skip OpenSSL
implementations when a strict crypto policy is active.

Use EVP_DigestInit_ex() instead of EVP_DigestInit() to initialize the
EVP context. The EVP_DigestInit() function clears alls flags and breaks
usedforsecurity flag again.

Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue38270

Automerge-Triggered-By: @tiran

### CODE REVIEW

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -649,20 +654,29 @@ def wrapper(*args, **kw):
...
     ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS
     ValueError: unsupported hash type md4
     """
     def decorator(func):
         @functools.wraps(func)
         def wrapper(*args, **kwargs):
             try:
-                hashlib.new(digestname)
+                if openssl:
+                    if _hashlib is None:
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: I would prefer to avoid having to inspect the _hashlib module to check which implementation picked hashlib. But it's not possible right now, so it's ok to rely on the private _hashlib module. My concern was that other Python implementation may want to reuse Python test suite, but may not have a _hashlib module.

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -649,20 +654,29 @@ def wrapper(*args, **kw):
...
     policy.
 
+    If 'openssl' is True, then the decorator checks that OpenSSL provides
+    the algorithm. Otherwise the check falls back to built-in
+    implementations. The usedforsecurity flag is passed to the constructor.
+
     ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS
     ValueError: unsupported hash type md4
     """
     def decorator(func):
```
Comment (vstinner) [inline Lib/test/support/__init__.py:670]: "import test.support" became an import nightmare: it imports way too many modules :-(

Would it be possible to import _hashlib here?

```
try:
    import _hashlib
except ImportError:
    _hashlib = None
```

File: Lib/test/test_hashlib.py [hunk 3]
```diff
@@ -403,6 +415,7 @@ def test_blocksize_name_blake2(self):
         self.check_blocksize_name('blake2b', 128, 64)
         self.check_blocksize_name('blake2s', 64, 32)
 
+
```
Comment (vstinner) [inline Lib/test/test_hashlib.py]: nitpick: PEP 8 suggests only 1 empty line between functions, you may revert this unrelated change.

**Review by encukou — APPROVED** [review]
This looks good to me. Please address Victor's comment (but I don't think the imports are that much of a problem).

File: Lib/test/support/__init__.py (hunk 2 above)
Comment (tiran) [reply Lib/test/support/__init__.py:670]: It's possible but it doesn't make much sense IMO. The function would have to try/import/except for every decorator call. The support helper package is already a kitchen sink.

Comment (vstinner) [reply Lib/test/support/__init__.py:670]: > It's possible but it doesn't make much sense IMO.

"import test.support" is used by regrtest, our test runner, and almost all tests. Reducing imports should make tests faster and more "isolated".

> The function would have to try/import/except for every decorator call. The support helper package is already a kitchen sink.

Packages are put in sys.modules and so only the first import is slow.

We should split test.support monster into subpackages like test.support.sslutils.
